### ISSUE TITLE
Use `User-Agent` header to have analytics

### ISSUE THREAD

**Issue by curquiza** [meilisearch/integration-guides#150]
Replaces: https://github.com/meilisearch/integration-guides/issues/58

Currently, we have no analytics about integrations.
As a first step, we would like to know the proportion of each integration.

Thanks to the [new spec implementation](https://github.com/meilisearch/specifications/blob/develop/text/0034-telemetry-policies.md), the `User-Agent` header will be sent to Segment (and then to Amplitude) via the MeiliSearch server 🎉 

Here is the expected format:
```
User-Agent: Meilisearch <language> (<version>)
```

-> The version should be in parenthesis and the information must start with `Meilisearch`

Ex:

```
User-Agent: Meilisearch JS (v0.22.1)
```

If multiple integrations are used during the user path, they should be separated by `;`.

```
User-Agent: Meilisearch Rails (v0.2.3) ; Meilisearch Ruby (v0.16.0)
```

-> In this example, it means the Ruby SDK should be able to accept a custom `User-Agent` that the Rails plugin can pass when initializing the Ruby client.

---------

### First step: do it basically in all the SDKs

The SDKs should send the expected `User-Agent` header. In this step, no need to implement a way to pass a custom `User-Agent` header.
We only need to find a smart way (if possible) to pass the version, and/or to check the version is well updated for each release (using the `.github/script/check-release.sh` script for example)

- [x] meilisearch-dotnet (https://github.com/meilisearch/meilisearch-dotnet/pull/241)
- [x] meilisearch-dart (https://github.com/meilisearch/meilisearch-dart/pull/129)
- [x] meilisearch-go (https://github.com/meilisearch/meilisearch-go/pull/279)
- [x] meilisearch-java (https://github.com/meilisearch/meilisearch-java/pull/503)
- [x] meilisearch-js (https://github.com/meilisearch/meilisearch-js/pull/1272)
- [x] meilisearch-php (https://github.com/meilisearch/meilisearch-php/pull/294) and https://github.com/meilisearch/meilisearch-php/pull/369
- [x] meilisearch-python (https://github.com/meilisearch/meilisearch-python/pull/433)
- [x] meilisearch-ruby (https://github.com/meilisearch/meilisearch-ruby/pull/292) and (https://github.com/meilisearch/meilisearch-ruby/pull/354)
- [x] meilisearch-rust (https://github.com/meilisearch/meilisearch-rust/pull/254)
- [x] meilisearch-swift (https://github.com/meilisearch/meilisearch-swift/pull/277)
- [x] meilisearch-strapi-plugin (https://github.com/meilisearch/strapi-plugin-meilisearch/pull/455)
- [x] docs-searchbar.js (https://github.com/meilisearch/docs-searchbar.js/pull/621)
- [x] meilisearch-firebase (https://github.com/meilisearch/firestore-meilisearch/pull/91)
- [x] meilisearch-rails (https://github.com/meilisearch/meilisearch-rails/pull/179)
- [x] meilisearch-symfony (https://github.com/meilisearch/meilisearch-symfony/pull/193)
- [x] meilisearch-vuepress (https://github.com/meilisearch/vuepress-plugin-meilisearch/pull/244)
- [x] meilisearch-gatsby (https://github.com/meilisearch/gatsby-plugin-meilisearch/pull/147)
- [x] instant-meilisearch (https://github.com/meilisearch/instant-meilisearch/pull/780)
- [x] laravel scout (https://github.com/laravel/scout/pull/681)
- [x] docs-scraper (https://github.com/meilisearch/docs-scraper/pull/387)

### Second step: do it in every integration

This step involves some SDKs (JS, PHP, Ruby, and probably more in the future) that can accept a custom `User-Agent` header. This needs an internal discussion for implementation and, why not, might deserve another issue.

**Comment by curquiza** [comment]
FYI @meilisearch/product-team

**Comment by gmourier** [comment]
Hi @bidoubiwa! Thank you for this complete and accurate feedback.

If I understand correctly, it's mostly languages running on Chrome that won't be able to send their User-Agent using the conforming header.

The solution could be to ask for a custom user-agent, we would like to avoid it for several reasons. It is possible that some HTTP clients will encounter problems by not being able to override the user-agent. It is also possible that some proxies will remove a custom header by default.

In the vast majority of cases, I have the impression that we won't have the information about the use of js SDKs via chrome. Am I wrong?

If so, we can make the assumption that it is the most used. There is no real need to know if it should be maintained or prioritized in my opinion but you tell me.

The idea behind this, if I'm not mistaken @curquiza, was to know to what extent we should prioritize efforts on SDKs that probably wouldn't be used.

**Comment by curquiza** [comment]
After the discussion with @bidoubiwa and @gmourier 

We keep `User-Agent` header for the moment to forward the SDK information into MeiliSearch. Each SDK should fill the `User-Agent` header.
The other integrations should be done in a second time since it will involve more implementation in the code base (cf the current issue)

The problem on Chrome we have for front-end is still present and should be solved in the future, for example by sending the analytics information as query parameters. We need first to ensure the first implementation we've done works.

TLDR;
The current issue description is still what we need to implement, nothing has changed, except we are more aware of some limits.

**Comment by bidoubiwa** [comment]
I agree with the proposal. Even without the information from chrome we still at least receive some information :) 
Just to be sure, the separation with the `;` it should be prefixed and sufixed with a whitespace ?

**Comment by gmourier** [comment]
> I agree with the proposal. Even without the information from chrome we still at least receive some information :) Just to be sure, the separation with the `;` it should be prefixed and sufixed with a whitespace ?

No need to separate them with a space. On the core side, we separate content-type with `;`. @irevoire do you confirm? It is rather for readability needs when inspecting HTTP headers with the naked eye.

**Comment by irevoire** [comment]
Yes we split the user-agent on the `;` and we put everything in a Set :+1:

**Comment by brunoocasali** [comment]
List of the qualified names used in the headers:

- `Meilisearch .NET`
- `Meilisearch PHP`
- `Meilisearch Ruby`
- `Meilisearch Dart`
- `Meilisearch JavaScript`
- `Meilisearch Rust`
- `Meilisearch Java`
- `Meilisearch Go`
- `Meilisearch Swift`

And:

- `Meilisearch Rails`
- `Meilisearch Symfony`

**Comment by gmourier** [comment]
Hi @brunoocasali

It seems it's missing `Meilisearch Python` here. Am I right?

Also, what's about Laravel Scout, Firebase, and Gatsby plugin? Could be nice to know it also!

**Comment by brunoocasali** [comment]
> Hi @brunoocasali! Thanks for the list!
> 
> It seems it's missing `Meilisearch Python` here. Am I right?
> 
> Also, what's about Laravel Scout, Firebase, and Gatsby plugins? Could be nice to know it also!

Oh yeah, I forgot to add them thanks for the reminder. About the Laravel, I'm not sure if we will be able to add the customization there.

**Comment by bidoubiwa** [comment]
I'm adding the missing docs-scraper repo

**Comment by brunoocasali** [comment]
We fuck*n made it, team. You're the best! 🎉🎉🎉
